File: packages/burkaz/lib/src/query.dart
import 'schema_field_options.dart';

import 'term.dart';

typedef Score = double;

/// A query on a burkaz index.
sealed class Query {
  const Query();
}

/// A query that matches all documents.
class AllQuery extends Query {
  const AllQuery();
}

/// A query that matches no documents.
class EmptyQuery extends Query {
  const EmptyQuery();
}

/// A query that matches a term.
class TermQuery extends Query {
  const TermQuery({
    required this.term,
    this.indexingStrategy = IndexingStrategy.basic,
  });

  /// The term to match.
  final Term term;

  /// The indexing strategy to use.
  final IndexingStrategy indexingStrategy;
}

/// A query that matches a set of terms.
class TermSetQuery extends Query {
  const TermSetQuery({required this.terms});

  /// The terms to match.
  final List<Term> terms;
}

/// A query that matches a fuzzy term.
class FuzzyTermQuery extends Query {
  const FuzzyTermQuery({
    required this.term,
    required this.distance,
    required this.transpositionCostOne,
    this.prefix = false,
  });

  /// The term to match.
  final Term term;

  /// The distance to match.
  final int distance;

  /// Whether to match the transposition cost one.
  final bool transpositionCostOne;

  /// Whether to match the prefix.
  final bool prefix;
}

/// A query that matches a phrase.
class PhaseQuery extends Query {
  const PhaseQuery({required this.terms, this.slop = 0});

  /// The terms to match.
  final List<Term> terms;

  /// The slop to match.
  final int slop;
}

/// A query that matches a regular expression phrase.
class RegexPhaseQuery extends Query {
  const RegexPhaseQuery({
    required this.field,
    required this.terms,
    this.slop = 0,
    this.maxExpansions = 1 << 14,
  });

  /// The field to match.
  final String field;

  /// The terms to match.
  final List<String> terms;

  /// The slop to match.
  final int slop;

  /// The max expansions to match.
  final int maxExpansions;
}

/// A query that matches all the child queries.
class AndQuery extends Query {
  const AndQuery(this.queries);

  /// The child queries to match.
  final List<Query> queries;
}

/// A query that matches at least one of the child queries.
class OrQuery extends Query {
  const OrQuery(this.queries);

  /// The child queries to match.
  final List<Query> queries;
}

/// A query that negates the child query.
class NotQuery extends Query {
  const NotQuery(this.query);

  /// The child query to negate.
  final Query query;
}

/// A query that boosts the score of the query.
class BoostQuery extends Query {
  const BoostQuery({required this.query, required this.boost});

  /// The child query to boost.
  final Query query;

  /// The boost to apply.
  final Score boost;
}

File: packages/burkaz_core/src/query.rs
use tantivy::query::{
    AllQuery, BooleanQuery, BoostQuery, EmptyQuery, FuzzyTermQuery, Occur, PhraseQuery, Query,
    RegexPhraseQuery, TermQuery, TermSetQuery,
};

use crate::{index::BurkazIndex, schema::BurkazIndexingStrategy, term::BurkazTerm};

/// A query that can be executed against the index.
pub enum BurkazQuery {
    /// Matches all documents.
    All,
    /// Matches no documents.
    Empty,
    /// All child queries must match (AND).
    And(Vec<BurkazQuery>),
    /// At least one child query must match (OR).
    Or(Vec<BurkazQuery>),
    /// The child query must not match (NOT).
    Not(Box<BurkazQuery>),
    /// Matches a term.
    Term {
        term: BurkazTerm,
        indexing_strategy: BurkazIndexingStrategy,
    },
    /// Matches a set of terms.
    TermSet {
        terms: Vec<BurkazTerm>,
    },
    /// Matches a fuzzy term.
    FuzzyTerm {
        term: BurkazTerm,
        distance: u8,
        transposition_cost_one: bool,
        prefix: bool,
    },
    /// Matches a phrase.
    Phase {
        terms: Vec<BurkazTerm>,
        slop: u32,
    },
    RegexPhase {
        field_id: u32,
        terms: Vec<String>,
        slop: u32,
        max_expansions: u32,
    },
    Boost {
        query: Box<BurkazQuery>,
        boost: f32,
    },
    Parse {
        query_text: String,
    },
}

impl BurkazQuery {
    pub unsafe fn from_raw(ptr: *mut Self) -> Self {
        unsafe { *Box::from_raw(ptr) }
    }

    pub fn into_raw(self) -> *mut Self {
        Box::into_raw(Box::new(self))
    }
}

impl BurkazQuery {
    pub fn to_tantivy_query(&self, index: &BurkazIndex) -> Box<dyn Query> {
        match self {
            Self::All => Box::new(AllQuery),
            Self::Empty => Box::new(EmptyQuery),
            Self::And(queries) => Box::new(BooleanQuery::new(
                queries
                    .iter()
                    .map(|query| (Occur::Must, query.to_tantivy_query(index)))
                    .collect(),
            )),
            Self::Or(queries) => Box::new(BooleanQuery::new(
                queries
                    .iter()
                    .map(|query| (Occur::Should, query.to_tantivy_query(index)))
                    .collect(),
            )),
            Self::Not(query) => Box::new(BooleanQuery::new(vec![(
                Occur::MustNot,
                query.to_tantivy_query(index),
            )])),
            Self::Term {
                term,
                indexing_strategy,
            } => Box::new(TermQuery::new(
                term.to_tantivy_term(),
                (*indexing_strategy).into(),
            )),
            Self::TermSet { terms } => Box::new(TermSetQuery::new(
                terms
                    .iter()
                    .map(|term| term.to_tantivy_term())
                    .collect::<Vec<_>>(),
            )),
            Self::FuzzyTerm {
                term,
                distance,
                transposition_cost_one,
                prefix,
            } => {
                if *prefix {
                    Box::new(FuzzyTermQuery::new_prefix(
                        term.to_tantivy_term(),
                        *distance,
                        *transposition_cost_one,
                    ))
                } else {
                    Box::new(FuzzyTermQuery::new(
                        term.to_tantivy_term(),
                        *distance,
                        *transposition_cost_one,
                    ))
                }
            }
            Self::Phase { terms, slop } => Box::new(PhraseQuery::new_with_offset_and_slop(
                terms
                    .iter()
                    .enumerate()
                    .map(|(index, term)| (index, term.to_tantivy_term()))
                    .collect(),
                *slop,
            )),
            Self::RegexPhase {
                field_id,
                terms,
                slop,
                max_expansions,
            } => {
                let mut query = Box::new(RegexPhraseQuery::new_with_offset_and_slop(
                    tantivy::schema::Field::from_field_id(*field_id),
                    terms.iter().cloned().enumerate().collect(),
                    *slop,
                ));
                query.set_max_expansions(*max_expansions);
                query
            }
            Self::Boost { query, boost } => {
                Box::new(BoostQuery::new(query.to_tantivy_query(index), *boost))
            }
            Self::Parse { query_text } => {
                let parsed_query = index
                    .query_parser()
                    .parse_query(&query_text)
                    .unwrap_or(Box::new(EmptyQuery));

                parsed_query
            }
        }
    }
}
